code/test/sixsq/nuvla/server/resources/voucher_lifecycle_test.clj

```diff
@@ -0,0 +1,223 @@
+            (ltu/body->edn)
+            (ltu/is-status 403))
+
+        ;; check expire acls - fail as anon
+        (-> session-anon
+            (request expire-url
+                     :request-method :post)
+            (ltu/body->edn)
+            (ltu/is-status 403))
+
+        ;; success as regular user
+        (-> session-user
+            (request activate-url
+                     :request-method :post)
+            (ltu/body->edn)
+            (ltu/is-status 200))
+
+        ;; now the redeem url should finally exist
+        (let [voucher-updated (-> session-admin
+                                  (request admin-abs-uri)
+                                  (ltu/body->edn)
+                                  (ltu/is-status 200)
+                                  (ltu/is-operation-present "redeem"))
+              redeem-url (str p/service-context (ltu/get-op voucher-updated "redeem"))]
+
+          ;; state is activated but anon cannot redeem, anon doesn't have can-edit
+          (-> session-anon
+              (request redeem-url
+                       :request-method :post)
+              (ltu/body->edn)
+              (ltu/is-status 403))
+
+          ; but user can redeem
+          (-> session-user
+              (request redeem-url
+                       :request-method :post)
+              (ltu/body->edn)
+              (ltu/is-status 200)))
+
+        ;; can expire it
+        (-> session-user
+            (request expire-url
+                     :request-method :post)
+            (ltu/body->edn)
+            (ltu/is-status 200))
+
+        ;; but cannot expire if already expired
+        (-> session-user
+            (request expire-url
+                     :request-method :post)
+            (ltu/body->edn)
+            (ltu/is-status 400))
+
+        ;; verify that an edit works
+        (let [updated (assoc voucher :target-audience "scientists@university.com")]
+
+          (-> session-admin
+              (request admin-abs-uri
+                       :request-method :put
+                       :body (json/write-str updated))
+              (ltu/body->edn)
+              (ltu/is-status 200)
+              :response
+              :body)
+
+          (let [updated-body (-> session-admin
+                                 (request admin-abs-uri)
+                                 (ltu/body->edn)
+                                 (ltu/is-status 200)
+                                 :response
+                                 :body)]
+
+            (is (= "scientists@university.com" (:target-audience updated-body))))))
+
+      ;; admin can delete the voucher
+      (-> session-admin
+          (request admin-abs-uri :request-method :delete)
+          (ltu/body->edn)
+          (ltu/is-status 200))
+
+      ;; user cannot delete the voucher
+      (-> session-user
+          (request user-abs-uri :request-method :delete)
+          (ltu/body->edn)
+          (ltu/is-status 403)))))
```

Comment:
I would have expected a 404 here instead. The admin just deleted the resource.  It shouldn't exist anymore, so there should be no ACL to compare to.

Reply:
in fact it should be 200, cause the uri is different, that's a 2nd resource